File: android/app/src/main/java/com/hcjeong/forestix/ui/screens/LocaleSetupSheet.kt
// First-run Country → Region cascade (internationalization framework).
// Replaces the US-only region picker: the cruiser picks a Country first, then
// — only for countries that have sub-national regions (the US) — a Region.
// The volume standard and, for the US, the default log rule are derived from
// the selection automatically (US West → Scribner, East → Doyle); metric
// countries carry no log rule.
//
// Shown once after install (and re-openable from Settings → Country & region).
// Dismissing without finishing is allowed — `regionPickerSeen = true` is
// stamped on any selection / skip / dismiss so the sheet never auto-presents
// again. Named `LocaleSetupSheet` to match the iOS screen of the same name
// (the persisted `regionPickerSeen` flag keeps its name across both platforms).
//
// Host it in a ModalBottomSheet from the caller:
//   if (!settings.regionPickerSeen) {
//       ModalBottomSheet(onDismissRequest = { ... }) { LocaleSetupSheet(onDismiss = { ... }) }
//   }

package com.hcjeong.forestix.ui.screens

import androidx.compose.foundation.border
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.List
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Eco
import androidx.compose.material.icons.filled.Landscape
import androidx.compose.material.icons.filled.Park
import androidx.compose.material.icons.filled.Public
import androidx.compose.material.icons.filled.WaterDrop
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.hcjeong.forestix.LocalAppEnvironment
import com.hcjeong.forestix.common.Country
import com.hcjeong.forestix.common.Region
import com.hcjeong.forestix.common.defaultLogRule
import com.hcjeong.forestix.ui.clickableNoRipple
import com.hcjeong.forestix.ui.theme.Forestix
import com.hcjeong.forestix.ui.theme.ForestixRadius
import com.hcjeong.forestix.ui.theme.ForestixSpace

@Composable
fun LocaleSetupSheet(onDismiss: () -> Unit) {
    val env = LocalAppEnvironment.current
    val settings by env.settings.state.collectAsStateWithLifecycle()
    val colors = Forestix.colors
    val type = Forestix.type

    // Two-step cascade: null = country step; a US country = region step.
    var regionStepCountry by remember { mutableStateOf<Country?>(null) }

    Column(
        Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = ForestixSpace.md)
            .padding(top = ForestixSpace.sm, bottom = ForestixSpace.xl),
        verticalArrangement = Arrangement.spacedBy(ForestixSpace.lg),
    ) {
        val inRegionStep = regionStepCountry != null

        // MARK: - Title bar (Back on the region step; Skip always)
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            if (inRegionStep) {
                TextButton(onClick = { regionStepCountry = null }) {
                    Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back",
                        modifier = Modifier.size(16.dp))
                    Text("  Country")
                }
            }
            Text(
                if (inRegionStep) "Pick your region" else "Where do you cruise?",
                style = type.bodyBold, color = colors.textPrimary,
                modifier = Modifier.weight(1f))
            TextButton(onClick = {
                env.settings.setRegionPickerSeen(true)
                onDismiss()
            }) {
                Text("Skip")
            }
        }

        if (!inRegionStep) {
            // MARK: - Country step
            Column(verticalArrangement = Arrangement.spacedBy(ForestixSpace.xs)) {
                Text("Which country?", style = type.bodyBold, color = colors.textPrimary)
                Text(
                    "Sets the species presets and the volume standard (board-foot " +
                        "log rules for the US, cubic metres elsewhere). You can change " +
                        "this any time in Settings.",
                    style = type.caption, color = colors.textSecondary)
            }
            Column(verticalArrangement = Arrangement.spacedBy(ForestixSpace.sm)) {
                Country.entries.forEach { c ->
                    PickerRow(
                        glyph = Icons.Filled.Public,
                        title = c.displayName,
                        subtitle = c.subtitle,
                        isSelected = settings.country == c && !inRegionStep,
                        trailingChevron = c.hasRegions,
                    ) {
                        env.settings.setCountry(c)
                        if (c.hasRegions) {
                            // Advance to the region step (US).
                            regionStepCountry = c
                        } else {
                            // Metric country: no region, no log rule — done.
                            env.settings.setRegion(null)
                            env.settings.setRegionPickerSeen(true)
                            onDismiss()
                        }
                    }
                }
            }
        } else {
            // MARK: - Region step (US only)
            val selectedRegion = Region.fromRaw(settings.region)
            Column(verticalArrangement = Arrangement.spacedBy(ForestixSpace.xs)) {
                Text("Which forest do you cruise?", style = type.bodyBold,
                    color = colors.textPrimary)
                Text(
                    "Pre-loads the right species and sets the default log rule for " +
                        "your region.",
                    style = type.caption, color = colors.textSecondary)
            }
            Column(verticalArrangement = Arrangement.spacedBy(ForestixSpace.sm)) {
                Region.entries.forEach { r ->
                    PickerRow(
                        glyph = glyph(r),
                        title = r.displayName,
                        subtitle = r.subtitle,
                        isSelected = selectedRegion == r,
                        trailingChevron = false,
                    ) {
                        env.settings.setRegion(r.raw)
                        // Derive the US default log rule from the region.
                        env.settings.setLogRule(r.defaultLogRule)
                        env.settings.setRegionPickerSeen(true)
                        onDismiss()
                    }
                }
            }
        }
    }
}

// MARK: - Row

@Composable
private fun PickerRow(
    glyph: ImageVector,
    title: String,
    subtitle: String,
    isSelected: Boolean,
    trailingChevron: Boolean,
    action: () -> Unit,
) {
    val colors = Forestix.colors
    val type = Forestix.type
    Surface(
        color = colors.surface,
        shape = ForestixRadius.card,
        modifier = Modifier
            .fillMaxWidth()
            .clip(ForestixRadius.card)
            .border(
                width = if (isSelected) 1.5.dp else 0.5.dp,
                color = if (isSelected) colors.primary else colors.divider,
                shape = ForestixRadius.card,
            )
            .clickableNoRipple(action),
    ) {
        Row(
            Modifier.fillMaxWidth().padding(ForestixSpace.md),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(ForestixSpace.md),
        ) {
            Box(
                Modifier
                    .size(36.dp)
                    .background(colors.primaryMuted, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Icon(glyph, contentDescription = null,
                    tint = colors.primary, modifier = Modifier.size(15.dp))
            }
            Column(
                Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                Text(title, style = type.bodyBold, color = colors.textPrimary)
                Text(subtitle, style = type.caption, color = colors.textSecondary,
                    maxLines = 2)
            }
            if (isSelected) {
                Icon(Icons.Filled.CheckCircle, contentDescription = "Selected",
                    tint = colors.primary, modifier = Modifier.size(17.dp))
            } else if (trailingChevron) {
                Icon(Icons.Filled.ChevronRight, contentDescription = null,
                    tint = colors.textTertiary, modifier = Modifier.size(12.dp))
            }
        }
    }
}

/// SF Symbol → Material icon mapping for the US region glyphs
/// (tree / mountain.2 / sun.max / drop / leaf / list.bullet).
private fun glyph(region: Region): ImageVector = when (region) {
    Region.PNW_WEST, Region.PNW_EAST, Region.N_ROCKIES -> Icons.Filled.Park
    Region.N_SIERRA, Region.S_SIERRA -> Icons.Filled.Landscape
    Region.CA_COAST -> Icons.Filled.Park
    Region.SOUTHWEST -> Icons.Filled.WbSunny
    Region.COASTAL_PLAIN, Region.BOTTOMLAND -> Icons.Filled.WaterDrop
    Region.PIEDMONT, Region.APPALACHIAN -> Icons.Filled.Eco
    Region.ALL -> Icons.AutoMirrored.Filled.List
}

File: TimberCruisingApp/Screens/LocaleSetupSheet.swift
// First-run locale cascade — replaces the US-only RegionPickerSheet.
//
//   Step 1  "Where are you cruising?"  → pick a Country.
//   Step 2  "Pick your region"         → US only; metric countries skip it.
//
// The volume standard and (US) default log rule are derived automatically:
//   • US West regions → Scribner, US East regions → Doyle (existing convention)
//   • metric countries → their m³ standard, no log rule.
//
// Dismissing at any point (Skip, a row pick, or a swipe-down) stamps
// `regionPickerSeen` — the presenting screen does this in its onDismiss — so
// the cascade never auto-nags again. Re-selectable any time from
// Settings → Country & region.

import SwiftUI

public struct LocaleSetupSheet: View {

    @EnvironmentObject private var settings: AppSettings
    @Environment(\.dismiss) private var dismiss
    @State private var path = NavigationPath()

    public init() {}

    public var body: some View {
        NavigationStack(path: $path) {
            countryStep
                .navigationDestination(for: Country.self) { country in
                    regionStep(for: country)
                }
        }
    }

    // MARK: - Step 1 · Country

    private var countryStep: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: ForestixSpace.lg) {
                intro(
                    title: "Where are you cruising?",
                    detail: "Sets your units and volume standard. Change it any time in Settings.")
                VStack(spacing: ForestixSpace.sm) {
                    ForEach(Country.allCases) { country in
                        LocaleRow(
                            glyph: country.flag,
                            title: country.displayName,
                            subtitle: country.subtitle,
                            isSelected: settings.country == country,
                            showsChevron: country.hasRegions
                        ) {
                            choose(country)
                        }
                    }
                }
            }
            .padding(.horizontal, ForestixSpace.md)
            .padding(.top, ForestixSpace.sm)
            .padding(.bottom, ForestixSpace.xl)
        }
        .background(ForestixPalette.canvas.ignoresSafeArea())
        .navigationTitle("Country")
        #if os(iOS)
        .navigationBarTitleDisplayMode(.inline)
        #endif
        .toolbar {
            ToolbarItem(placement: .cancellationAction) {
                Button("Skip") {
                    settings.regionPickerSeen = true
                    dismiss()
                }
            }
        }
    }

    // MARK: - Step 2 · Region (US only)

    private func regionStep(for country: Country) -> some View {
        ScrollView {
            VStack(alignment: .leading, spacing: ForestixSpace.lg) {
                intro(
                    title: "Pick your region",
                    detail: "Loads the right species list and log rule for your region. Change it any time in Settings.")
                VStack(spacing: ForestixSpace.sm) {
                    ForEach(country.regions) { region in
                        LocaleRow(
                            glyph: nil,
                            systemGlyph: regionGlyph(region),
                            title: region.displayName,
                            subtitle: region.subtitle,
                            isSelected: settings.region == region,
                            showsChevron: false
                        ) {
                            choose(region)
                        }
                    }
                }
            }
            .padding(.horizontal, ForestixSpace.md)
            .padding(.top, ForestixSpace.sm)
            .padding(.bottom, ForestixSpace.xl)
        }
        .background(ForestixPalette.canvas.ignoresSafeArea())
        .navigationTitle("Pick your region")
        #if os(iOS)
        .navigationBarTitleDisplayMode(.inline)
        #endif
    }

    // MARK: - Selection

    private func choose(_ country: Country) {
        settings.country = country
        // Selecting a country drives the unit system (US → imperial, metric
        // countries → metric). The manual Units toggle in Settings can still
        // override this afterwards.
        settings.unitSystem = country.defaultUnitSystem
        if country.hasRegions {
            path.append(country)          // → region step
        } else {
            settings.region = nil          // metric: single implicit region
            settings.regionPickerSeen = true
            dismiss()
        }
    }

    private func choose(_ region: Region) {
        settings.region = region
        settings.logRule = region.defaultLogRule   // West→Scribner, East→Doyle
        settings.regionPickerSeen = true
        dismiss()
    }

    // MARK: - Bits

    private func intro(title: String, detail: String) -> some View {
        VStack(alignment: .leading, spacing: ForestixSpace.xs) {
            Text(title)
                .font(ForestixType.bodyBold)
                .foregroundStyle(ForestixPalette.textPrimary)
            Text(detail)
                .font(ForestixType.caption)
                .foregroundStyle(ForestixPalette.textSecondary)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    private func regionGlyph(_ region: Region) -> String {
        switch region {
        case .pnwWest, .pnwEast, .nRockies: return "tree"
        case .nSierra, .sSierra:            return "mountain.2"
        case .caCoast:                      return "tree"
        case .southwest:                    return "sun.max"
        case .coastalPlain, .bottomland:    return "drop"
        case .piedmont, .appalachian:       return "leaf"
        case .all:                          return "list.bullet"
        }
    }
}

// MARK: - Row

private struct LocaleRow: View {
    var glyph: String? = nil          // emoji flag (country rows)
    var systemGlyph: String? = nil    // SF Symbol (region rows)
    let title: String
    let subtitle: String
    let isSelected: Bool
    let showsChevron: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: ForestixSpace.md) {
                ZStack {
                    Circle()
                        .fill(ForestixPalette.primaryMuted)
                        .frame(width: 36, height: 36)
                    if let glyph {
                        Text(glyph).font(.system(size: 18))
                    } else if let systemGlyph {
                        Image(systemName: systemGlyph)
                            .font(.system(size: 15, weight: .medium))
                            .foregroundStyle(ForestixPalette.primary)
                    }
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(ForestixType.bodyBold)
                        .foregroundStyle(ForestixPalette.textPrimary)
                    Text(subtitle)
                        .font(ForestixType.caption)
                        .foregroundStyle(ForestixPalette.textSecondary)
                        .lineLimit(2)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer(minLength: 0)
                if isSelected {
                    Image(systemName: "checkmark.circle.fill")
                        .foregroundStyle(ForestixPalette.primary)
                } else if showsChevron {
                    Image(systemName: "chevron.forward")
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(ForestixPalette.textTertiary)
                }
            }
            .padding(ForestixSpace.md)
            .frame(maxWidth: .infinity)
            .background(
                RoundedRectangle(cornerRadius: ForestixRadius.card, style: .continuous)
                    .fill(ForestixPalette.surface))
            .overlay(
                RoundedRectangle(cornerRadius: ForestixRadius.card, style: .continuous)
                    .stroke(isSelected ? ForestixPalette.primary : ForestixPalette.divider,
                            lineWidth: isSelected ? 1.5 : 0.5))
        }
        .buttonStyle(.plain)
    }
}
